src/chain/forkChoice/attestationAggregator.ts

```diff
@@ -0,0 +1,110 @@
+import BN from "bn.js";
+
+import {Gwei, Slot, ValidatorIndex} from "../../types";
+
+
+/**
+ * Root is a block root as a hex string
+ * Used here for light weight and easy comparison
+ */
+export type Root = string;
+
+/**
+ * Minimal representation of attsetation for the purposes of fork choice
+ */
+export interface ForkChoiceAttestation {
+  target: Root;
+  attester: ValidatorIndex;
+  weight: Gwei;
+}
+
+/**
+ * Attestation aggregated across participants
+ */
+export interface AggregatedAttestation {
+  target: Root;
+  weight: Gwei;
+  prevWeight: Gwei;
+}
+
+/**
+ * Keep track of the latest attestations per validator
+ * as well as aggregated attestations per block root
+ */
+export class AttestationAggregator {
+
+  /**
+   * aggregation: target -> sum of all attestations
+   */
+  public latestAggregates: Record<Root, AggregatedAttestation>;
+
+  /**
+   * lookup: validator -> target + weight contributed by validator
+   */
+  public latestAttestations: Record<ValidatorIndex, ForkChoiceAttestation>;
+
+  private slotLookup: (Root) => Slot | null;
+
+  public constructor(slotLookup: (Root) => Slot | null) {
+    this.latestAggregates = {};
+    this.latestAttestations = {}
+    this.slotLookup = slotLookup;
+  }
+
+  private ensureAggregate(target: Root): AggregatedAttestation {
+    if (!this.latestAggregates[target]) {
+      this.latestAggregates[target] = {
+        target,
+        weight: new BN(0),
+        prevWeight: new BN(0),
+      };
+    }
+    return this.latestAggregates[target];
+  }
+
+  /**
+   * Add an attestion into the aggregator
+   */
+  public addAttestation(a: ForkChoiceAttestation): void {
+    const prevA = this.latestAttestations[a.attester];
+    if (prevA) {
+      // Previous individual attestation exists
+      const prevSlot = this.slotLookup(prevA.target);
+      const newSlot = this.slotLookup(a.target);
+      if (prevSlot === null || newSlot === null || prevSlot > newSlot) {
+        // new attestation is not new enough or slot doesn't exist
+        return;
+      }
+      const newAgg = this.ensureAggregate(a.target);
+      const prevAgg = this.latestAggregates[prevA.target];
+      if (prevA.target !== a.target) { // new attestation target
```

Comment:
I believe the conditional should be `prevA.target !== newAgg`

Reply:
I think this is is right as is. `prevA` is the previous attestation for a given validator, `a` is the current incoming attestation for the same validator.
If the targets are not the same, then `a` is a new attestation for a different block that should replace the old one.

Reply:
Based on Proto's implementation you should be making the comparison to `newAgg` because you need to ensure that the aggregate exists in the first place. `newAgg` is `target`, except fro the fact that if there was no aggregate already, it adds one

Reply:
thoughts @wemeetagain ?

Reply:
The aggregate targets will always match the underlying attestation targets because they are derived from the attestation targets.
And as long as any attester attests a target, the aggregation for that target will exist (will not be pruned).
 
What we need to check here is if the previous attestation target is different than the current incoming attestation target. This can be done by looking at the derived aggregate attestations, or it can be done by comparing the attestations directly.

This is equivalent to proto's code, but I think more readable, because we're comparing the what we actually care about, the attestation targets, rather than the object equality of the prevAgg and newAgg, which happens as a result of the attestation targets being equal.

It turns out though, that object equality is marginally faster than string equality. So I'll update to `newAgg !== prevAgg`.